compiler-builtins: plumb LSE support for aarch64 on linux/gnu when optimized-compiler-builtins not enabled

pmur opened the pull request:
Add dynamic support for aarch64 LSE atomic ops on linux/gnu targets when optimized-compiler-builtins is not enabled.

Enabling LSE is the primary motivator for rust-lang/rust#143689, though extending the rust version doesn't seem too farfetched. Are there more details which I have overlooked which make this impractical? I've tested this on an aarch64 host with LSE.

r? @tgross35 

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

pmur commented on library/compiler-builtins/compiler-builtins/src/aarch64_linux.rs:
```diff
@@ -24,7 +21,38 @@
-// We don't do runtime dispatch so we don't have to worry about the `__aarch64_have_lse_atomics` global ctor.
+/// non-zero if the host supports LSE atomics.
+#[cfg(not(feature = "c"))]
+static mut HAVE_LSE_ATOMICS: u8 = 0;
```
Ok, that works here too.

pmur commented on library/compiler-builtins/compiler-builtins/src/aarch64_linux.rs:
```diff
@@ -24,7 +21,38 @@
+/// should also be OK as the fallback atomic operations always work.
+#[used]
+#[unsafe(link_section = ".init_array.90")]
+#[cfg(all(not(feature = "c"), target_env = "gnu"))]
```
outline-atomics are only enabled on the gnu targets, is there need beyond that? Likewise, the initializer will get included without a gate for `c` when linking compiler-rt. It should be harmless, but isn't needed.

pmur committed:
Dynamically enable LSE for aarch64 rust provided intrinsics

Create a private module to hold the bootstrap code needed enable LSE
at startup on aarch64-*-linux-* targets when rust implements the
intrinsics.

This is a bit more heavyweight than compiler-rt's LSE initialization,
but has the benefit of initializing the aarch64 cpu feature detection
for other uses.

Using the rust initialization code does use some atomic operations,
that's OK. Mixing LSE and non-LSE operations should work while the
update flag propagates.

pmur commented on library/compiler-builtins/compiler-builtins/src/aarch64_linux.rs:
```diff
@@ -150,6 +219,7 @@ macro_rules! compare_and_swap {
                     "cbnz   w17, 0b",
                     "1:",
                     "ret",
+                    sym crate::aarch64_linux::HAVE_LSE_ATOMICS,
```
Oh, thanks. That's really helpful and not well described in the inline asm docs.

Amanieu commented:
At the moment this is the only way I can see this working. Fundamentally compiler-builtins is not able to call into other libraries so LSE must be manually initialized from outside.

pmur commented on library/compiler-builtins/compiler-builtins/src/aarch64_linux.rs:
```diff
@@ -24,7 +21,15 @@
+/// Call to enable LSE atomic operations.
+pub fn enable_lse() {
+    HAVE_LSE_ATOMICS.store(1, Ordering::Relaxed);
+}
```
Ok, I'll update the usage. I'll take this out of draft when I push the update.

pmur committed:
compiler-builtins: plumb LSE support for aarch64 on linux

Add dynamic support for aarch64 LSE atomic ops on linux targets
when optimized-compiler-builtins is not enabled.

A hook, __enable_rust_lse, is provided for the runtime to enable
them if available. A future patch will use this to enable them
if available.

The resulting asm should exactly match that of LLVM's compiler-rt
builtins, though the symbol naming for the support function and
global does not.

tgross35 reviewed (approved):
Thanks for the updates; this looks good to me, but I'd like Amanieu to take a look as well.

If https://github.com/rust-lang/rust/pull/144761 doesn't land before this, I'll add a commit there to update the config here.

r? @Amanieu

Amanieu commented on library/std/src/sys/configure_builtins.rs:
```diff
@@ -0,0 +1,27 @@
+#[cfg(all(
+    target_arch = "aarch64",
+    target_os = "linux",
+    any(target_env = "gnu", target_env = "musl"),
```
You don't need to check target-env at all here. Any linux target should be fine.

pmur replied:
Ok, removed.

Amanieu commented on library/compiler-builtins/compiler-builtins/src/aarch64_linux.rs:
```diff
@@ -137,6 +178,11 @@ macro_rules! compare_and_swap {
             ) -> int_ty!($bytes) {
                 // We can't use `AtomicI8::compare_and_swap`; we *are* compare_and_swap.
                 core::arch::naked_asm! {
+                    jmp_if_no_lse!("8f"),
```
It would be better if this macro instead took the LSE instruction as an argument and included the `ret` and `8:`. This would reduce the code here to a single line.

pmur replied:
Ok, updated.

taiki-e commented on library/compiler-builtins/compiler-builtins/src/aarch64_linux.rs:
```diff
@@ -126,6 +135,41 @@ macro_rules! stxp {
+macro_rules! try_lse_op {
+    ($op: literal, $ordering:ident, $bytes:tt, $($reg:literal,)* [ $mem:ident ] ) => {
+        concat!(
+            ".arch_extension lse; ",
```
AFAIK, there is currently no reliable way to use .arch_extension in naked_asm/global_asm correctly in general cases due to the LLVM bug. (See https://github.com/rust-lang/rust/pull/137720#discussion_r2014505753 and https://github.com/rust-lang/rust/pull/137720#discussion_r1973608259 for more info.) 
Although maybe there is some special rule for compiler-builtins...

pmur replied:
From experimentation, it seems like `.arch_extension` only applied to the function. Specifying the raw instruction as `.inst 0x...` without toggling features also works, but is ugly. If that's preferable, I'll update the PR.

Amanieu replied:
In the context of compiler-builtins, just using `.arch_extension` is probably fine since it is always compiled separately and is never included in LTO.
